Ansible: Add missing cucumber tests and fix issue on SLS

meaksh opened the pull request:
## What does this PR change?

This PR implements missing cucumber tests for Ansible feature in normal and SSH minion.  

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: **tests**

- [x] **DONE**

## Test coverage
- Cucumber tests were added

- [x] **DONE**

## Links

Tracks: https://github.com/SUSE/spacewalk/issues/14644

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [x] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

meaksh commented:
Unassigning QA people since this PR is targetting `ansible-integration` atm.

srbarrios commented on testsuite/features/secondary/min_ansible_control_node.feature:
```diff
@@ -4,6 +4,9 @@
 @scope_ansible
 Feature: Operate an Ansible control node in a normal minion
 
+  Scenario: Pre-requisite: Deploy test playbooks and inventory file
```
I noticed that the full feature gives 3 whitespaces instead of 2 in the steps of the scenario.
Could you fix it, please?

meaksh replied:
Okk :+1:

srbarrios commented on testsuite/features/secondary/min_ansible_control_node.feature:
```diff
@@ -12,31 +15,62 @@ Feature: Operate an Ansible control node in a normal minion
+     When I follow "Ansible" in the content area
+     Then I should see a "Ansible Control Node Configuration" text
+     And I enter "/srv/playbooks/" as "new_playbook_path_input"
+     And I click on button with id "new_playbook_path_save"
```
Could we use the text of the button instead of the ID?
It's more readable for a user following these steps.

meaksh replied:
No, I think we cannot. 

Here we have two different "Save" buttons (with same text) in the same page, and the `<div>` where they are contained have no ids to distinguise between them. So I cannot reuse one of the existing step and had to implement this new one.

srbarrios commented on testsuite/features/secondary/min_ansible_control_node.feature:
```diff
@@ -12,31 +15,62 @@ Feature: Operate an Ansible control node in a normal minion
+     When I follow "Ansible" in the content area
+     And I follow "Inventories" in the content area
+     And I wait until I see "/srv/playbooks/example_playbook2_orion_dummy/hosts" text
+     Then I click on "/srv/playbooks/example_playbook2_orion_dummy/hosts" text in Ansible paths
```
I don't care too much about this, but there are team members that could not like using a "Then" here :)

meaksh replied:
Got it

srbarrios commented on testsuite/features/secondary/min_ansible_control_node.feature:
```diff
@@ -12,31 +15,62 @@ Feature: Operate an Ansible control node in a normal minion
+     And I wait until I see "/srv/playbooks" text
+     And I click on "/srv/playbooks" text in Ansible paths
+     And I wait until I see "/srv/playbooks/example_playbook2_orion_dummy/example_playbook2_orion_dummy.yml" text
+     Then I click on "example_playbook2_orion_dummy/example_playbook2_orion_dummy.yml"
```
So we use "When" for all the actions triggered to the system provoking a "reaction", and "Then" when we expect some result from this action.

meaksh replied:
Thanks for clarifying!

srbarrios commented on testsuite/features/step_definitions/common_steps.rb:
```diff
@@ -1292,6 +1292,9 @@ def server_secret
   source = File.dirname(__FILE__) + '/../upload_files/ansible/playbooks/example_playbook2_orion_dummy/hosts'
   return_code = file_inject(target, source, dest + "hosts")
   raise 'File injection failed' unless return_code.zero?
+  source = File.dirname(__FILE__) + '/../upload_files/ansible/playbooks/example_playbook2_orion_dummy/example_file.txt'
```
What about having this other structure?
```
.
└── ansible
    └── playbooks
        ├── orion_dummy
        │   ├── file.txt
        │   ├── hosts
        │   └── playbook_orion_dummy.yml
        └── ping
            └── playbook_ping.yml
```

meaksh replied:
Yeah, It makes sense to simplify this a bit. In any case, I'm keeping `playbook_ping.yml` in the main `playbooks` as I don't want that one to be contained in a custom folder as the `orion_dummy` one.

srbarrios commented on testsuite/features/step_definitions/navigation_steps.rb:
```diff
@@ -244,6 +244,21 @@
   end
 end
 
+#
+# Click on a button with a given "id"
+When(/^I click on button with id "([^"]*)"$/) do |button_id|
+  find(:xpath, ".//button[@id=\"#{button_id}\"]").click
+end
+
```
You don't need a special step to click by Id.

The step `I click on "zzz"` will check in both text and ids.

![image](https://user-images.githubusercontent.com/2827771/116702388-4ff66300-a9c9-11eb-83ca-bc52fd2fcd70.png)

srbarrios replied:
In any case, as I commented in another part of the PR, I strongly recommend to search by text and not by ID.
As it's easier to follow the steps by someone that reads the .feaure file.

meaksh replied:
Ahhh! I can pass the `id` directly to the "I click on" step.

Ok. I'm fixing this - Thanks!

srbarrios commented on testsuite/features/step_definitions/navigation_steps.rb:
```diff
@@ -244,6 +244,21 @@
+  find(:xpath, ".//button[@id=\"#{button_id}\"]").click
+end
+
+#
+# Click on a text which appears inside of <div> with
+# the given "id"
+When(/^I click on "([^"]*)" text in Ansible paths$/) do |text|
+  within(:xpath, '//div[@id="ansible-path-content"]') do
+    find('h6', text: text).click
+  end
+end
+
```
I would say we don't need this special step.

Could you instead try to use this step in your scenario?

``` #
# Click on a button which appears inside of <div> with
# the given "id"
When(/^I click on "([^"]*)" in element "([^"]*)"$/) do |text, element_id|
  within(:xpath, "//div[@id=\"#{element_id}\"]") do
    click_button_and_wait(text, match: :first)
  end
end
```

meaksh replied:
I tried that one but it didn't work since it's expecting a button.

In this case, what we have here is not a button but a "collapsible" div that it's only containing `<h6>text</h6>` where I need to click to get the content loaded.

That's why I had to implement this particular step to deal with those "paths" boxes in the Ansible UI.

srbarrios commented on testsuite/features/secondary/min_ansible_control_node.feature:
```diff
@@ -12,31 +15,62 @@ Feature: Operate an Ansible control node in a normal minion
+     When I follow "Ansible" in the content area
+     And I follow "Inventories" in the content area
+     And I wait until I see "/srv/playbooks/example_playbook2_orion_dummy/hosts" text
+     Then I click on "/srv/playbooks/example_playbook2_orion_dummy/hosts" text in Ansible paths
```
Try to use this step instead:
`I click on "/srv/playbooks/example_playbook2_orion_dummy/hosts" in element "ansible-path-content"`

Not sure if it will work, due to the extra filter 'h6' you added, but it's worth trying it, and safe an additional step.

meaksh replied:
As mentioned above, that didn't work for me

ktsamis reviewed (approved):
Approving to get some runs over the weekend, we can take another look Monday and adjust as needed.

meaksh commented:
@srbarrios I addressed your comments and tackled some of your suggestions. I'm merging this already to have testsuite executions by the weekend. 

Let's talk on Monday in case there is something else to be tackled here. Thanks!
